### PR TITLE
Decote 2015

### PR DESCRIPTION
**Pull request by adrienpacifico**
Import reforms/plf2015.py into hard coded legislation. Tests for the reform not implemented.

### CODE REVIEW

File: openfisca_france/model/prelevements_obligatoires/impot_revenu/ir.py [hunk 1]
```diff
@@ -1254,18 +1254,31 @@ def function(self, simulation, period):
...
+    @dated_function(start = date(2001, 1, 1), stop = date(2013, 12, 31))
+    def function_2001_2013(self, simulation, period):
         period = period.start.offset('first-of', 'year').period('year')
         ir_plaf_qf = simulation.calculate('ir_plaf_qf', period)
         decote = simulation.legislation_at(period.start).ir.decote
 
         return period, (ir_plaf_qf < decote.seuil) * (decote.seuil - ir_plaf_qf) * 0.5
 
+    @dated_function(start = date(2014, 1, 1))
+    def function_2014(self, simulation, period):
```
Comment (benjello) [inline openfisca_france/model/prelevements_obligatoires/impot_revenu/ir.py]: SHould be: function_2014__(....

File: openfisca_france/tests/base.py [hunk 2]
```diff
@@ -31,7 +31,6 @@
     aides_ville_paris,
     allocations_familiales_imposables,
     cesthra_invalidee,
-    plf2015,
```
Comment (benjello) [inline openfisca_france/tests/base.py]: Keep this test. It is used for revenus 2013.

File: openfisca_france/model/prelevements_obligatoires/impot_revenu/ir.py (hunk 1 above)
Comment (adrienpacifico) [inline openfisca_france/model/prelevements_obligatoires/impot_revenu/ir.py]: Ok, I'll correct and add a .yaml test in my next push.

Comment (adrienpacifico) [inline openfisca_france/model/prelevements_obligatoires/impot_revenu/ir.py]: Is there official guidelines that I missed for that ?

File: openfisca_france/tests/base.py (hunk 2 above)
Comment (adrienpacifico) [inline openfisca_france/tests/base.py]: Ok

File: openfisca_france/model/prelevements_obligatoires/impot_revenu/ir.py (hunk 1 above)
Comment (benjello) [inline openfisca_france/model/prelevements_obligatoires/impot_revenu/ir.py]: Not written: first _ for separatio nand another _ for unknow en date

**Comment by adrienpacifico** [comment]
Wait for travis (that should be ok), but is it ok now ?

File: openfisca_france/model/prelevements_obligatoires/impot_revenu/ir.py (hunk 1 above)
Comment (adrienpacifico) [inline openfisca_france/model/prelevements_obligatoires/impot_revenu/ir.py]: Thx, got it. Some IRPP date are given with a yearMonthDay notation, since fiscal rules apply usually only over a fiscal year, should't we change that to startYear_endYear notation ?

File: openfisca_france/tests/base.py (hunk 2 above)
Comment (adrienpacifico) [inline openfisca_france/tests/base.py]: I'm sorry but the only way to pass the test for me is to comment/erase those lines.

**Comment by benjello** [comment]
We should wait for [this] (https://github.com/openfisca/openfisca-france/pull/318) to be merged

**Comment by cbenz** [comment]
@benjello I've restarted Travis build after having merged #318

**Comment by cbenz** [comment]
`decote` became a `DatedFormulaColumn` with @adrienpacifico changes, but [`decote` in `.py`](https://github.com/openfisca/openfisca-france/blob/c363f2fff502a5519e1685b61de4c60371d5f4aa/openfisca_france/reforms/plf2015.py#L42-L55) remained a `SimpleFormulaColumn`.

I think this is why the tests fail.

@benjello @adrienpacifico Should the reform be updated?

**Comment by benjello** [comment]
@cbenz @adrienpacifico : the reform should be updated

File: openfisca_france/reforms/plf2015.py [hunk 3]
```diff
@@ -26,23 +26,24 @@
 from __future__ import division
 
 from openfisca_core import formulas, periods, reforms
-
+from ..model.base import *
```
Comment (adrienpacifico) [inline openfisca_france/reforms/plf2015.py]: Might be a bit overkill ?

File: openfisca_france/reforms/plf2015.py [hunk 4]
```diff
@@ -26,23 +26,24 @@
...
 
     @Reform.formula
-    class decote(formulas.SimpleFormulaColumn):
+    class decote(formulas.DatedFormulaColumn):
         label = u"Nouvelle décote 2015"
         reference = ir.decote
 
-        def function(self, simulation, period):
+        @dated_function(start = date(2013, 1, 1), stop = date(2013, 12, 31) )
+        def function_2013(self, simulation, period):
```
Comment (adrienpacifico) [inline openfisca_france/reforms/plf2015.py]: @benjello since it is only for 2013 is the notation ok ?
Plus, this implementation of the reform might not allow to use reform_tax_benfit_system for any other year than 2013 ?
@cbenz does a Dated formula reform modify just the dated function concerned or does it modify all the class (and thus you end with just one dated_function implemented in your reform) ?

File: openfisca_france/tests/formulas/irpp.yaml [hunk 5]
```diff
@@ -952,3 +952,20 @@
...
     irpp: -48036
+- name: "IRPP - Célibataire ayant des 	Revenu net imposable de 16000 € (teste la décote)"
+  period: 2014
+  absolute_error_margin: 0.9 #TODO : check why the error margin is so large, is there bottom round made by dgfip ?
+  input_variables:
+    salaire_imposable: 16000
+  output_variables:
+    irpp: -183
+    decote: 476
+#TODO : check where the 24.5 difference in IRPP comes from.
```
Comment (adrienpacifico) [inline openfisca_france/tests/formulas/irpp.yaml]: I've got something like 25 euros difference in that test compare to the DgFip simulator, that's not huge. Since your are not supposed to commit a test that do not pass, I commented it. @benjello Is it the way we should do ?

File: openfisca_france/reforms/plf2015.py (hunk 4 above)
Comment (benjello) [inline openfisca_france/reforms/plf2015.py]: It is intentionnal. You can change the label to Nouvelle décote 2015 sur revenus 2013.
When PLF2015 (revenus 2014) is announced we have only the 2013 legislation so we do a reform on that year.
When we have the full 2014 legislation we implement it and we shoud also implement a 2014 business as usual (not done yet) so you can plau your reform on 2013 and 2014.

File: openfisca_france/tests/formulas/irpp.yaml (hunk 5 above)
Comment (benjello) [inline openfisca_france/tests/formulas/irpp.yaml]: The PR cannot be accepted if we do not have the irrp right. It is the main point of the tests.
You should investigate. Be careful a lot of legislation parameters are not updated for 2014.
I am working on it using the ipp baremes.

Comment (adrienpacifico) [inline openfisca_france/tests/formulas/irpp.yaml]: Ok then I'll wait and rebase till parameters haven't been actualized. I'm not sure that I want to investigate on that difference for now since I do not have any real clue. 

If you have a specific task to do or a specific point to check, I'm willing to do it.

Comment (benjello) [inline openfisca_france/tests/formulas/irpp.yaml]: min/max abbattements sur les salaires et seuils des tranches de l'IR

File: openfisca_france/reforms/plf2015.py (hunk 4 above)
Comment (cbenz) [inline openfisca_france/reforms/plf2015.py]: > @cbenz does a Dated formula reform modify just the dated function concerned or does it modify all the class (and thus you end with just one dated_function implemented in your reform) ?

@adrienpacifico good question, I'll investigate with @eraviart (cf openfisca/openfisca-core#346)

**Comment by cbenz** [comment]
Tests are OK, should i merge?

File: openfisca_france/tests/formulas/irpp.yaml (hunk 5 above)
Comment (adrienpacifico) [inline openfisca_france/tests/formulas/irpp.yaml]: seuil de tranches de l'ir 2015 : C'est moi qui les avais mis en décembre, je viens de re-checker c'est ok.

min/max abbattements sur les salaire je ne sais pas ce que c'est je checke.

**Comment by adrienpacifico** [comment]
@benjello doesn't want to because I found a slight difference between DgFip simulator and OF (look at  commented lines).

File: openfisca_france/tests/formulas/irpp.yaml (hunk 5 above)
Comment (adrienpacifico) [inline openfisca_france/tests/formulas/irpp.yaml]: min/max abattements sur les salaires, tu as raison j'ai actualisé les nouvelles valeurs.

Comment (benjello) [inline openfisca_france/tests/formulas/irpp.yaml]: Ca fonctionne mieux ?

Comment (adrienpacifico) [inline openfisca_france/tests/formulas/irpp.yaml]: The tests pass with new min/max abatements, I push for a fresh PR.

Comment (adrienpacifico) [inline openfisca_france/tests/formulas/irpp.yaml]: @benjello Oui c'est ok, bravo pour ta perspicacité à trouver le problème. J'attend le travis et vous pouvez merger si tout est ok pour toi.

File: openfisca_france/param/param.xml [hunk 6]
```diff
@@ -2781,7 +2784,16 @@
         <VALUE deb="2010-01-01" fin="2011-12-31" valeur="878" />
         <VALUE deb="2012-01-01" fin="2012-12-31" valeur="960" />
         <VALUE deb="2013-01-01" fin="2013-12-31" valeur="1016" />
-        <VALUE deb="2014-01-01" fin="2014-12-31" valeur="1021" />
```
Comment (benjello) [inline openfisca_france/param/param.xml]: @adrienpacifico: why did you remove this line : is it wrong ?

Comment (adrienpacifico) [inline openfisca_france/param/param.xml]: Well except if I'm totally off track the new decote threshold has two parameters in 2014 one for singles and one for couples. That parameters might have been entered before the implementation of the reform (i'll check with gitk... )

Comment (adrienpacifico) [inline openfisca_france/param/param.xml]: It's been added by @clbe on 2014/10/30, I don't remember but probably before the decote reform and suppression of the first tax slice has been announced.

Comment (benjello) [inline openfisca_france/param/param.xml]: Reforms where not in good shape and it was difficult to modify the
parameters. So we changed it in the code and the trick was done

Comment (adrienpacifico) [inline openfisca_france/param/param.xml]: So everything is ok ?

Comment (benjello) [inline openfisca_france/param/param.xml]: I thought you changed the abattement sur les pensions

**Comment by iDevelopper** [comment]
La décote a changé en 2015 (revenus de 2014)

Pour un couple 1870 €, pour un célibataire 1135 €.

Salaire imposable pour un célibataire : 15.000 € -> droits issus du barème : 743 €
Décote : 1 135 € - 743 € = 392 €
Impôt : 351 €

**Comment by benjello** [comment]
@adrienpacifico could you write a test with @iDevelopper test ?

**Comment by iDevelopper** [comment]
Salaire imposable pour un couple : 30.000 € -> droits issus du barème : 1487 €
Décote : 1 870 € - 1 487 € = 383 €
Impôt : 1104 €

**Comment by adrienpacifico** [comment]
@iDevelopper I am sorry but I do not find the same results on the official tax simulator : http://www3.finances.gouv.fr/calcul_impot/2015/simplifie/index.htm

L'impôt pour le célibataire à 15.000 euros est de 0 décote 533 euros
L'impôt pour le couple à 30.000 euros est de 264 euros pour 803 euros de décote

@benjello 
Pour un célibataire à 16000 euros le test est comité dans ma PR et passe déjà.
Pour un couple je ne sais pas comment faire et mon essai de test foire, sans doc je ne peux pas implémenter un nouveau test pour l'IR.

Par ailleurs la décote ne pourra pas être testée avec les résultats donnée par le site de la DgFip voir : [#326](https://github.com/openfisca/openfisca-france/issues/326)

**Comment by adrienpacifico** [comment]
Test for couple a 30k euros is done in irpp.yaml see last push.

File: openfisca_france/param/param.xml (hunk 6 above)
Comment (adrienpacifico) [inline openfisca_france/param/param.xml]: Nope added a line before, and suppress a line 600 lines after.

Comment (iDevelopper) [inline openfisca_france/param/param.xml]: Adrien,

Célibataire, 1AJ =18.000 (revenus d’activités) et 1AK = 3.000 (frais réels)
-> Salaire imposable = 15.000

Patrick BODET

De :  adrienpacifico <notifications@github.com>
Répondre à :  openfisca/openfisca-france
<reply+001b32b238c70f86adc012b9ceee7c9961ece61e67e140d192cf0000000112198aa89
2a170ce026498b4@reply.github.com>
Date :  mardi 22 septembre 2015 23:31
À :  openfisca/openfisca-france <openfisca-france@noreply.github.com>
Cc :  Patrick BODET <patrick.bodet4@wanadoo.fr>
Objet :  Re: [openfisca-france] Decote 2015 (#324)


In openfisca_france/param/param.xml
<https://github.com/openfisca/openfisca-france/pull/324#discussion_r40147124
> :
> @@ -2781,7 +2784,16 @@
>          <VALUE deb="2010-01-01" fin="2011-12-31" valeur="878" />
>          <VALUE deb="2012-01-01" fin="2012-12-31" valeur="960" />
>          <VALUE deb="2013-01-01" fin="2013-12-31" valeur="1016" />
> -        <VALUE deb="2014-01-01" fin="2014-12-31" valeur="1021" />
Nope added a line before, and suppress a line 600 lines after.

—
Reply to this email directly or view it on GitHub
<https://github.com/openfisca/openfisca-france/pull/324/files#r40147124> .

Comment (iDevelopper) [inline openfisca_france/param/param.xml]: Adrien,

Documentation :

http://bofip.impots.gouv.fr/bofip/2495-PGP

Patrick BODET

De :  adrienpacifico <notifications@github.com>
Répondre à :  openfisca/openfisca-france
<reply+001b32b238c70f86adc012b9ceee7c9961ece61e67e140d192cf0000000112198aa89
2a170ce026498b4@reply.github.com>
Date :  mardi 22 septembre 2015 23:31
À :  openfisca/openfisca-france <openfisca-france@noreply.github.com>
Cc :  Patrick BODET <patrick.bodet4@wanadoo.fr>
Objet :  Re: [openfisca-france] Decote 2015 (#324)


In openfisca_france/param/param.xml
<https://github.com/openfisca/openfisca-france/pull/324#discussion_r40147124
> :
> @@ -2781,7 +2784,16 @@
>          <VALUE deb="2010-01-01" fin="2011-12-31" valeur="878" />
>          <VALUE deb="2012-01-01" fin="2012-12-31" valeur="960" />
>          <VALUE deb="2013-01-01" fin="2013-12-31" valeur="1016" />
> -        <VALUE deb="2014-01-01" fin="2014-12-31" valeur="1021" />
Nope added a line before, and suppress a line 600 lines after.

—
Reply to this email directly or view it on GitHub
<https://github.com/openfisca/openfisca-france/pull/324/files#r40147124> .

File: openfisca_france/tests/formulas/irpp.yaml (hunk 5 above)
Comment (cbenz) [inline openfisca_france/tests/formulas/irpp.yaml]: > Since your are not supposed to commit a test that do not pass, I commented it.

@adrienpacifico Thanks to ask!

Tests must be valid in order to be merged into `next`, but since you are on your branch you can leave invalid tests.

For me these invalid tests an indicator that I must not merge the pull-request :)

Once you fix them, the pull-request will launch Travis tests again automatically. Did you know that if you push new commits in the pull-request branch (ie `decote_2015`), the pull-request updates itself?

Comment (adrienpacifico) [inline openfisca_france/tests/formulas/irpp.yaml]: @cbenz  thank you for your answer.
>Tests must be valid in order to be merged into next, but since you are on your branch you can leave invalid tests.

@benjello  @cbenz, ok since those not passing test are new test that I added and do not break old test I taught that it was the best way to do. It also gives an incentive to hide old errors to the guy that do the pull request if one is entitled to solve any error he finds. 

>Once you fix them, the pull-request will launch Travis tests again automatically. Did you know that >if you push new commits in the pull-request branch (ie decote_2015), the pull-request updates >itself?

Yes I know I've done that 4 times only for this pull request...

>For me these invalid tests an indicator that I must not merge the pull-request :)

For current PR we've found with @benjello  the problem with my commented test (old parameters), and I also added an other test for couples. Except for style (pep8) issue or mistake found I will not push again for this new PR :-) .

**Comment by adrienpacifico** [comment]
@benjello @cbenz I think now you can merge if you want to.

**Comment by benjello** [comment]
Good to merge @cbenz @adrienpacifico

File: openfisca_france/param/param.xml (hunk 6 above)
Comment (adrienpacifico) [inline openfisca_france/param/param.xml]: @iDevelopper thanx. If you just fill 1AJ (which is net imposable on your fiche de paye) you have to substract 10% as an rebate on your business expenses. So most people will have to put 16666 (15000*0.9) euros to get your exemple. Feel free to implement it in a .yaml test and commit it :-) !

Comment (iDevelopper) [inline openfisca_france/param/param.xml]: Sur le simulateur des impôts, j’ai saisi 18000 € dans AJ et 3000 €  de frais
réels dans AK : Cela donne un revenu imposable de 15000 € et une décote de
392 €. Mais on ne parle pas de la même chose tous les deux, toi tu parles du
net de la fiche de paie qui n’est pas le revenu imposable mais le net fiscal
à déclarer et moi je parle du revenu imposable calculé par le fisc en
prenant en compte notamment soit la déduction forfaitaire des 10% ou des
frais réels.

Patrick BODET

De :  adrienpacifico <notifications@github.com>
Répondre à :  openfisca/openfisca-france
<reply+001b32b27265ec8ddb570683285aae3c7b200b03805d2bec92cf00000001121a2a8c9
2a170ce026520aa@reply.github.com>
Date :  mercredi 23 septembre 2015 10:53
À :  openfisca/openfisca-france <openfisca-france@noreply.github.com>
Cc :  Patrick BODET <patrick.bodet4@wanadoo.fr>
Objet :  Re: [openfisca-france] Decote 2015 (#324)


In openfisca_france/param/param.xml
<https://github.com/openfisca/openfisca-france/pull/324#discussion_r40181930
> :
> @@ -2781,7 +2784,16 @@
>          <VALUE deb="2010-01-01" fin="2011-12-31" valeur="878" />
>          <VALUE deb="2012-01-01" fin="2012-12-31" valeur="960" />
>          <VALUE deb="2013-01-01" fin="2013-12-31" valeur="1016" />
> -        <VALUE deb="2014-01-01" fin="2014-12-31" valeur="1021" />
@iDevelopper <https://github.com/iDevelopper>  thanx. If you just fill 1AJ
(which is net imposable on your fiche de paye) you have to substract 10% as
an rebate on your business expenses. So most people will have to put 16666
(15000*0.9) euros to get your exemple. Feel free to implement it in a .yaml
test and commit it :-) !

—
Reply to this email directly or view it on GitHub
<https://github.com/openfisca/openfisca-france/pull/324/files#r40181930> .

**Comment by cbenz** [comment]
Great! Thanks @iDevelopper @adrienpacifico @benjello
